### PR TITLE
Rework signal handling

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
### Rework signal handling to handle transactions as they are received

Currently, transactions are accumulated till the block state is accepted and the accumulated transactions are sent along with the block state to be handled. This is a problem because when a user calls the send_transaction api, it will take till the speculative block is aborted and the next block is applied before all the transactions are then recorded. This means that if the user tries to immediately get the status of the transaction, it will get back “UNKNOWN” till the transaction is finally processed (up to 500 ms later).

### Integration test work

Added integration tests to verify calls to get_transaction_status for a known transaction progressed from "APPLIED_LOCALLY" to "IN_BLOCK" to "IRREVERSIBLE" and then, after passing the success duration, returning "UNKNOWN". Also added a test to verify that get_transaction_status reported the correct status when a transaction's block was forked out.

### Issue Resolution

Also resolves Issue: https://github.com/eosnetworkfoundation/mandel/issues/176

```
curl -sd '{"id": "240ed958f05b672f9d15bdb332a49fa9e53715ec18272649b951d8f91bfb23e7"}' http://localhost:8888/v1/chain/get_transaction_status | jq
{
  "state": "IN_BLOCK",
  "block_number": 84,
  "block_id": "0000005492ae08787b46c6cc07852fc9391be887dd79e980bcd9245a574c9f56",
  "block_timestamp": "2022-05-06T15:34:17.500",
  "expiration": "2022-05-06T15:34:47.000",
  "head_number": 326,
  "head_id": "00000146f7df7ece3967526f35bcafcf5ae090fbfca208910f20eab73a03efa6",
  "head_timestamp": "2022-05-06T15:36:18.500",
  "irreversible_number": 23,
  "irreversible_id": "0000001727b6d94facf760a323a943ccfe9249224fecce23c402c1f66d09df20",
  "irreversible_timestamp": "2022-05-06T15:33:47.000",
  "earliest_tracked_block_id": "000000110b2a85d157470bfbba56802db4bbad616c0959027d27bd18cdb0d9f1",
  "earliest_tracked_block_number": 17
}

```

```
curl -sd '{"id": "00"}' http://localhost:8888/v1/chain/get_transaction_status | jq
{
  "code": 400,
  "message": "Invalid Request",
  "error": {
    "code": 3200006,
    "name": "invalid_http_request",
    "what": "invalid http request",
    "details": [
      {
        "message": "Invalid transaction id",
        "file": "chain_api_plugin.cpp",
        "line_number": 62,
        "method": "parse_params<eosio::chain_apis::read_only::get_transaction_status_params>"
      }
    ]
  }
}
```

```
curl -sd '' http://localhost:8888/v1/chain/get_transaction_status | jq
{
  "code": 400,
  "message": "Invalid Request",
  "error": {
    "code": 3200006,
    "name": "invalid_http_request",
    "what": "invalid http request",
    "details": [
      {
        "message": "A Request body is required",
        "file": "chain_api_plugin.cpp",
        "line_number": 54,
        "method": "parse_params<eosio::chain_apis::read_only::get_transaction_status_params>"
      }
    ]
  }
}

```

### CODE REVIEW

File: plugins/chain_interface/include/eosio/chain/signals_processor.hpp [hunk 1]
```diff
@@ -11,66 +11,60 @@ namespace eosio::chain {
 class signals_processor {
```
Comment (heifner) [inline plugins/chain_interface/include/eosio/chain/signals_processor.hpp:11]: I don't see this adding value anymore, I think it should just be removed and wire to the signals directly to trx finality status processor.

File: plugins/chain_plugin/test/test_trx_finality_status_processing.cpp [hunk 2]
```diff
@@ -839,12 +890,28 @@ namespace {
...
       }
 
       void send_spec_block() {
          status.signal_block_start(bn);
-         status.signal_applied_transactions(pre_block, no_bs);
+
+         for (const auto &trx_tuple : pre_block)
+         {
+            const auto &trace = std::get<0>(trx_tuple);
+            const auto &txn = std::get<1>(trx_tuple);
```
Comment (heifner) [inline plugins/chain_plugin/test/test_trx_finality_status_processing.cpp:911]: style

File: plugins/chain_interface/include/eosio/chain/signals_processor.hpp (hunk 1 above)
Comment (heifner) [reply plugins/chain_interface/include/eosio/chain/signals_processor.hpp:11]: Do make sure to move over the `FC_LOG_AND_DROP`

File: plugins/chain_plugin/trx_finality_status_processing.cpp [hunk 3]
```diff
@@ -40,82 +47,130 @@ namespace eosio::chain_apis {
...
+         handle_rollback();
+      }
+
+      const fc::time_point now = fc::time_point::now();
+      bool status_expiry = status_expiry_of_trxs(now);
+      if (status_expiry) {
+         determine_earliest_tracked_block_id();
+      }
+
+      // if this approve block was proceeded by speculative transactions that had
```
Comment (heifner) [inline plugins/chain_plugin/trx_finality_status_processing.cpp:154]: Update comment. `// if this approve block was proceeded by speculative transactions then we produced the block, update trx state.`

Comment (jgiszczak) [reply plugins/chain_plugin/trx_finality_status_processing.cpp:154]: And spelling.  I'm fairly sure you meant 'preceded'.

File: plugins/chain_plugin/test/test_trx_finality_status_processing.cpp (hunk 2 above)
Comment (oschwaldp-oci) [reply plugins/chain_plugin/test/test_trx_finality_status_processing.cpp]: Addressed: 3b7c2ec

File: plugins/chain_plugin/test/test_trx_finality_status_processing.cpp [hunk 4]
```diff
@@ -839,12 +890,28 @@ namespace {
 
       void send_block() {
          status.signal_block_start(bn);
-         status.signal_applied_transactions(block, bs);
+
+         for (const auto &trx_tuple : block)
+         {
+            const auto &trace = std::get<0>(trx_tuple);
+            const auto &txn = std::get<1>(trx_tuple);
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_plugin/test/test_trx_finality_status_processing.cpp]: Addressed: 3b7c2ec

File: plugins/chain_plugin/test/test_trx_finality_status_processing.cpp [hunk 5]
```diff
@@ -615,10 +651,23 @@ BOOST_AUTO_TEST_CASE(trx_finality_status_logic) { try {
    const auto bs_19_alt = make_block_state(bn);
    // const auto bs_19_alt = make_block_state(make_block_id(bn), std::vector<chain::packed_transaction_ptr>{});
    status.signal_block_start(bn);
-   status.signal_applied_transactions(trx_pairs_19_alt, bs_19_alt);
+
+   for (const auto &trx_tuple : trx_pairs_19_alt) {
+      const auto &trace = std::get<0>(trx_tuple);
+      const auto &txn = std::get<1>(trx_tuple);
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_plugin/test/test_trx_finality_status_processing.cpp]: Addressed: 3b7c2ec

File: plugins/chain_plugin/test/test_trx_finality_status_processing.cpp [hunk 6]
```diff
@@ -55,7 +55,7 @@ auto get_public_key( chain::name keyname, std::string role = "owner" ) {
    return get_private_key( keyname, role ).get_public_key();
 }
 
-auto make_unique_trx( const fc::microseconds& expiration ) {
+auto  make_unique_trx( const fc::microseconds& expiration ) {
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_plugin/test/test_trx_finality_status_processing.cpp]: Addressed: 3b7c2ec

File: plugins/chain_plugin/trx_finality_status_processing.cpp (hunk 3 above)
Comment (oschwaldp-oci) [reply plugins/chain_plugin/trx_finality_status_processing.cpp]: Addressed: 4399330

File: tests/trx_finality_status_forked_test.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+    if not nonProdNode.relaunch(nonProdNode.nodeNum):
+        errorExit("Failure - (non-production) node %d should have restarted" % (nonProdNode.nodeNum))
+
+    Print("Wait for LIB to move, which indicates prodNode[1] has forked out the branch")
+    assert cluster.waitOnClusterSync(blockAdvancing=1, blockType=BlockType.lib), \
+        Print("ERROR: Network did not reach concensus after bridge node was restarted.")
+
+    Print("Wait till prodNodes[1] is reporting at least the same head block number as the forked out block")
+    assert prodNodes[1].waitForBlock(retStatus["block_number"]), \
+        Print("Production node 1 should continue to advance, even after bridge node is killed. \n\nretStatus: {}".
```
Comment (jgiszczak) [inline tests/trx_finality_status_forked_test.py:298]: Assert message does not match the current expected state.  The bridge node has been relaunched by this point.

**Review by jgiszczak — CHANGES_REQUESTED** [review]
Submitting review prematurely because Github has decided to stop displaying my comments in line, for some reason.

I did stop calling out all individual instances of strings which should be converted to f-strings, and also stopped calling out individual instances of `assert` being used incorrectly.  All such changes should still be made.

File: tests/trx_finality_status_test.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            Print("ERROR: getTransactionStatus should contain \"expiration\" always.\nstatus: {}".
+                  format(json.dumps(status, indent=1)))
+
+        assert "earliest_tracked_block_id" in status and "earliest_tracked_block_number" in status, \
+            Print("ERROR: getTransactionStatus should contain \"earliest_tracked_block_id\" and \"earliest_tracked_block_number\" always.\nstatus: {}".
+                  format(json.dumps(status, indent=1)))
+
+    validate(status[0])
+
+    numTries = 5
```
Comment (jgiszczak) [inline tests/trx_finality_status_test.py:187]: Why not Node `waitForTransInBlock()` with a timeout of 5 * block time?  Or better yet 12 * block time, to guarantee the current producer finishes its turn and the next producer produces at least one block.

Generally speaking, when picking arbitrary delays, be cognizant of block times and the duration of a producer's turn (when there's more than one) to give the system the best chance to complete the test successfully (when not testing expected behavior which is fully deterministic).

File: tests/trx_finality_status_forked_test.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+#
+#  Test to verify that transaction finality status feature is
+#  working appropriately when forks occur
+#
+###############################################################
+Print=Utils.Print
+
+from core_symbol import CORE_SYMBOL
+
+def analyzeBPs(bps0, bps1, expectDivergence):
```
Comment (jgiszczak) [inline tests/trx_finality_status_forked_test.py:30]: Remove unused function.

**Review by jgiszczak — CHANGES_REQUESTED** [review]

File: plugins/chain_interface/include/eosio/chain/signals_processor.hpp (hunk 1 above)
Comment (oschwaldp-oci) [reply plugins/chain_interface/include/eosio/chain/signals_processor.hpp]: Addressed:  9d87fcbe4a4f2214c0260780d3d2a512d5326cbc

File: tests/trx_finality_status_forked_test.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+        Utils.errorExit(msg)
+
+    return firstDivergence
+
+def getMinHeadAndLib(prodNodes):
+    info0=prodNodes[0].getInfo(exitOnError=True)
+    info1=prodNodes[1].getInfo(exitOnError=True)
+    headBlockNum=min(int(info0["head_block_num"]),int(info1["head_block_num"]))
+    libNum=min(int(info0["last_irreversible_block_num"]), int(info1["last_irreversible_block_num"]))
+    return (headBlockNum, libNum)
```
Comment (oschwaldp-oci) [inline tests/trx_finality_status_forked_test.py:116]: Looks like this function is also unused.

File: tests/trx_finality_status_forked_test.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+        assert prodNodes[1].waitForNextBlock(), Print("Production node 1 should continue to advance, even after bridge node is killed")
+
+    postInfo = prodNodes[1].getInfo()
+    Print("preInfo: {}\n\npostInfo: {}".format(json.dumps(preInfo, indent=1), json.dumps(postInfo, indent=1)))
+
+    assert state == inBlockState, \
+        Print("ERROR: getTransactionStatus didn't return \"{}\" state.\n\nstatus: {}".format(inBlockState, json.dumps(retStatus, indent=1)))
+
+    Print("Relaunching the non-producing bridge node to connect the nodes")
+    if not nonProdNode.relaunch(nonProdNode.nodeNum):
```
Comment (oschwaldp-oci) [inline tests/trx_finality_status_forked_test.py:289]: Should not be passing any value to relaunch().  It relaunches the node being controlled by that instance.  It does not accept a node ID.

File: tests/trx_finality_status_forked_test.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    while numTries > 0:
+        retStatus = prodNodes[1].getTransactionStatus(transId)
+        state = getState(retStatus)
+        if state == inBlockState:
+            break
+        numTries -= 1
+        assert prodNodes[1].waitForNextBlock(), "Production node 1 should continue to advance, even after bridge node is killed"
+
+    postInfo = prodNodes[1].getInfo()
+    Print("preInfo: {json.dumps(preInfo, indent=1)}\n\npostInfo: {json.dumps(postInfo, indent=1)}")
```
Comment (oschwaldp-oci) [inline tests/trx_finality_status_forked_test.py:193]:
```suggestion
    Print(f"preInfo: {json.dumps(preInfo, indent=1)}\n\npostInfo: {json.dumps(postInfo, indent=1)}")
```

File: tests/trx_finality_status_test.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+    assert status[0]["head_number"] < status[1]["head_number"], \
+        f"ERROR: Successive calls to getTransactionStatus should return increasing values for \"head_number\".\n1st status: {json.dumps(status[0], indent=1)}\n\n2nd status: {json.dumps(status[1], indent=1)}"
+
+    block_number = status[1]["block_number"]
+    assert testNode.waitForIrreversibleBlock(block_number, timeout=180), \
+        f"ERROR: Failed to advance irreversible block to {block_number}. \nAPI Node info: {json.dumps(prod0.getInfo(), indent=1)}\n\nProducer info: {json.dumps(testNode.getInfo(), indent=1)}"
+
+    retStatus=testNode.getTransactionStatus(transId)
+    Print(f"retStatus: {retStatus}")
```
Comment (jgiszczak) [inline tests/trx_finality_status_test.py:200]: Debug output left over?

File: tests/trx_finality_status_test.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        f"\n1st status: {json.dumps(status[0], indent=1)}\n\n2nd status: {json.dumps(status[1], indent=1)}" + \
+        f"\n\nfinal status: {json.dumps(retStatus, indent=1)}"
+    validate(retStatus)
+
+    leeway=4
+    assert testNode.waitForBlock(blockNum=startingBlockNum+(successDuration*2),timeout=successDuration+leeway)
+
+    recentBlockNum=testNode.getBlockNum()
+    retStatus=testNode.getTransactionStatus(transId)
+    Print(f"retStatus: {retStatus}")
```
Comment (jgiszczak) [inline tests/trx_finality_status_test.py:213]: Debug output left over?  These don't seem particularly useful for the normal progress of the test.

**Review by jgiszczak — CHANGES_REQUESTED** [review]

File: tests/Node.py [hunk 15]
```diff
@@ -1389,6 +1399,19 @@ def isLibAdvancing():
             return self.getIrreversibleBlockNum() > currentLib
         return Utils.waitForBool(isLibAdvancing, timeout)
 
+    def waitForProducer(self, producer, timeout=None, exitOnError=False):
+        if timeout is None:
+            # default to the typical configuration of 21 producers, each producing 12 blocks in a row
+            timeout = 21 * 12;
```
Comment (jgiszczak) [inline tests/Node.py:1405]: Each producing 12 blocks in a row at half second intervals.  21 * 6.  `Utils.waitForBool()` expects seconds.

**Review by jgiszczak — CHANGES_REQUESTED** [review]

**Review by brianjohnson5972 — APPROVED** [review]

**Review by jgiszczak — APPROVED** [review]

**Pull request merged**